### PR TITLE
[PAC][CodeGen][ELF][AArch64] Support signed GOT with tiny code model

### LINKED ISSUE #87984: [PAC][ELF][lld] Support signed GOT

### PR DESCRIPTION
**Pull request by kovdan01**
Support the following relocations and assembly operators:

- `R_AARCH64_AUTH_GOT_ADR_PREL_LO21` (`:got_auth:` for `adr`)
- `R_AARCH64_AUTH_GOT_LD_PREL19` (`:got_auth:` for `ldr`)

`LOADgotAUTH` pseudo-instruction is expanded to actual instruction
sequence like the following.

```
adr x16, :got_auth:sym
ldr x0, [x16]
autia x0, x16
```

If FPAC bit is not set, a check+trap sequence similar to one used for `AUT`
pseudo is emitted to ensure auth success.

Both SelectionDAG and GlobalISel are suppported. For FastISel, we fall
back to SelectionDAG.

Tests starting with 'ptrauth-' have corresponding variants w/o this prefix.

### CODE REVIEW

**Comment by kovdan01** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113812?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#113812** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113812?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#113811** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113811?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @kovdan01 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp [hunk 1]
```diff
@@ -2277,28 +2277,40 @@ void AArch64AsmPrinter::LowerLOADgotAUTH(const MachineInstr &MI) {
   const MachineOperand &GAMO = MI.getOperand(1);
   assert(GAMO.getOffset() == 0);
 
-  MachineOperand GAHiOp(GAMO);
-  MachineOperand GALoOp(GAMO);
-  GAHiOp.addTargetFlag(AArch64II::MO_PAGE);
-  GALoOp.addTargetFlag(AArch64II::MO_PAGEOFF | AArch64II::MO_NC);
+  if (MI.getParent()->getParent()->getTarget().getCodeModel() ==
+      CodeModel::Tiny) {
```
Comment (ilovepi) [inline llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:2281]:
```suggestion
  if (MI.getMF()->getTarget().getCodeModel() == CodeModel::Tiny) {
```
Not sure if there's an easier way to get to the target from MI off the top of my head, but you can get the MachineFunction w/o going through `getParent`

**Review by ilovepi — COMMENTED** [review]
Overall, this seems fine as a translation of the spec. I still need to take another pass over the tests to be sure, but they seem appropriate so far. 

All that said, I'm far from an expert on Aarch64 minutiae, so a more experienced reviewer here will have to provide any final approval.

**Comment by kovdan01** [comment]
Sorry, I've accidentally pushed a wrong merge commit which cluttered the PR triggered a lot of reviewers.

Please ignore that PR, it'll be re-opened.

**Pull request closed without merging**